Mathlib/Topology/Connected/Clopen.lean

```diff
@@ -514,6 +522,125 @@ theorem continuous_coe : Continuous (mk : α → ConnectedComponents α) :=
 theorem range_coe : range (mk : α → ConnectedComponents α) = univ :=
   surjective_coe.range_eq
 
+lemma nonempty_iff_nonempty : Nonempty (ConnectedComponents α) ↔ Nonempty α :=
+  ⟨fun _ ↦ ConnectedComponents.surjective_coe.nonempty, fun h ↦ h.map ConnectedComponents.mk⟩
+
+instance [Nonempty α] : Nonempty (ConnectedComponents α) := by
+  rwa [ConnectedComponents.nonempty_iff_nonempty]
+
+lemma isEmpty_iff_isEmpty : IsEmpty (ConnectedComponents α) ↔ IsEmpty α := by
+  rw [← not_iff_not, not_isEmpty_iff, not_isEmpty_iff, nonempty_iff_nonempty]
+
+instance subsingleton [PreconnectedSpace α] : Subsingleton (ConnectedComponents α) := by
+  refine ⟨fun x y ↦ ?_⟩
+  obtain ⟨x, rfl⟩ := surjective_coe x
+  obtain ⟨y, rfl⟩ := surjective_coe y
+  simp_rw [coe_eq_coe, PreconnectedSpace.connectedComponent_eq_univ, ]
+
+section
+
+variable {ι : Type*} {U : ι → Set α} (hclopen : ∀ i, IsClopen (U i))
+  (hdisj : Pairwise (Disjoint on U)) (hunion : ⋃ i, U i = Set.univ)
+  (hconn : ∀ i, IsPreconnected (U i))
+
+include hclopen hdisj hunion in
+/-- A pairwise disjoint cover by clopens partitions the connected components. -/
+noncomputable def equivOfIsClopen : ConnectedComponents α ≃ Σ i, ConnectedComponents (U i) := by
```

Comment:
Perhaps add some lemmas describing this equiv? e.g. for `x ∈ U i`, `equivOfIsClopen (.mk x) = ⟨i, .mk ⟨x, _⟩⟩` and also the inverse